Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/SANS/SANSNormalizeToMonitor.py

```diff
@@ -0,0 +1,216 @@
+        # incident monitor.
+        incident_monitor_spectrum_number = normalize_to_monitor_state.incident_monitor
+        incident_monitor_spectrum_as_string = str(incident_monitor_spectrum_number)
+        background_tof_monitor_start = normalize_to_monitor_state.background_TOF_monitor_start
+        background_tof_monitor_stop = normalize_to_monitor_state.background_TOF_monitor_stop
+
+        if incident_monitor_spectrum_as_string in list(background_tof_monitor_start.keys()) and \
+           incident_monitor_spectrum_as_string in list(background_tof_monitor_stop.keys()):
+            start_tof = background_tof_monitor_start[incident_monitor_spectrum_as_string]
+            stop_tof = background_tof_monitor_stop[incident_monitor_spectrum_as_string]
+        else:
+            start_tof = normalize_to_monitor_state.background_TOF_general_start
+            stop_tof = normalize_to_monitor_state.background_TOF_general_stop
+
+        # Only if a TOF range was set, do we have to perform a correction
+        if start_tof and stop_tof:
+            flat_name = "CalculateFlatBackground"
+            flat_options = {"InputWorkspace": workspace,
+                            "OutputWorkspace": EMPTY_NAME,
+                            "StartX": start_tof,
+                            "EndX": stop_tof,
+                            "Mode": "Mean"}
+            flat_alg = create_unmanaged_algorithm(flat_name, **flat_options)
+            flat_alg.execute()
+            workspace = flat_alg.getProperty("OutputWorkspace").value
+        return workspace
+
+    def _convert_to_wavelength(self, workspace, normalize_to_monitor_state):
+        """
+        Converts the workspace from time-of-flight units to wavelength units
+
+        :param workspace: a time-of-flight workspace.
+        :param normalize_to_monitor_state: a SANSStateNormalizeToMonitor object.
+        :return: a wavelength workspace.
+        """
+        wavelength_low = normalize_to_monitor_state.wavelength_low
+        wavelength_high = normalize_to_monitor_state.wavelength_high
+        wavelength_step = normalize_to_monitor_state.wavelength_step
+        wavelength_step_type = normalize_to_monitor_state.wavelength_step_type
+        wavelength_rebin_mode = normalize_to_monitor_state.rebin_type
+        convert_name = "SANSConvertToWavelengthAndRebin"
+        convert_options = {"InputWorkspace": workspace,
+                           "WavelengthLow": wavelength_low,
+                           "WavelengthHigh": wavelength_high,
+                           "WavelengthStep": wavelength_step,
+                           "WavelengthStepType": RangeStepType.to_string(wavelength_step_type),
+                           "RebinMode": RebinType.to_string(wavelength_rebin_mode)}
+
+        convert_alg = create_unmanaged_algorithm(convert_name, **convert_options)
+        convert_alg.setPropertyValue("OutputWorkspace", EMPTY_NAME)
+        convert_alg.setProperty("OutputWorkspace", workspace)
```

Comment:
That is a good question.

To answer your second part first:  `convert_options` is a `dict` which gets passed into `create_unmanaged_algorithm` which then take the keys of the `dict` as the property name and the value of the `dict` as the property value and sets it on `alg.setProperty`. In the case above I need to set a string (see reason below) using the `alg.setPropertyValue`. I could have added have added `{"OutputWorkspace": workspace}` to the `convert_options, but since we have to set `OutputWorkspace` twice here, I wanted to leave it together.

Regarding your first question, the answer is slightly complex (but good to know). When Antii reviewed my last PR he had the same question. My reply was:

>The algorithms are run in general with unnamed workspaces, ie workspaces which are not placed on the ADS. You might have seen that the `OutputWorkspace` property needs to be set, something like
```python
...
alg.setProperty("InputWorkspace", ws)
...
alg.setProperty("OutputWorkspace", "test")
alg.execute()
out_ws = alg.getProperty("OutputWorkspace").value
....
```
>Mantid needs names for the `OutputWorkspace` property even if they are only temporary. It also means that output workspace is most likely a copy of the input. In my case here I want to perform an in-place operation on an unnamed workspace, hence I need to provide the workspace itself to the output and I need to provide a temporary name.

>The example below might illustrate this
```python
# -----------------------------------------------------------------------------------------------------------------------
# Create a fake algorithm which peforms some in-place operation
# -----------------------------------------------------------------------------------------------------------------------
class FakeAlgorithm(DataProcessorAlgorithm):
    def PyInit(self):
        self.declareProperty(MatrixWorkspaceProperty("InputWorkspace", '', optional=PropertyMode.Mandatory, direction=Direction.Input), doc='input')
    def PyExec(self):
        ws = self.getProperty("InputWorkspace").value
        
        alg = AlgorithmManager.createUnmanaged("ConvertUnits")
        alg.initialize()
        alg.setChild(True)
        alg.setProperty("InputWorkspace", ws)
        alg.setProperty("Target", "Wavelength")
        # UNCOMMENT LINE BELOW TO MAKE IT WORK
        #alg.setPropertyValue("OutputWorkspace", "EMPTY")
        alg.setProperty("OutputWorkspace", ws)
        
        alg.execute()
        ws_converted = alg.getProperty("OutputWorkspace").value

AlgorithmFactory.subscribe(FakeAlgorithm)


# -----------------------------------------------------------------------------------------------------------------------
# Provide an unnamed workspace
# -----------------------------------------------------------------------------------------------------------------------
alg_created = AlgorithmManager.createUnmanaged("CreateSampleWorkspace")
alg_created.initialize()
alg_created.setChild(True)
alg_created.setProperty("OutputWorkspace", "dummy")
alg_created.execute()
out_ws = alg_created.getProperty("OutputWorkspace").value


# -----------------------------------------------------------------------------------------------------------------------
# Run the algorithm
# -----------------------------------------------------------------------------------------------------------------------
alg_fake = AlgorithmManager.createUnmanaged("FakeAlgorithm")
alg_fake.initialize()
alg_fake.setChild(True)
alg_fake.setProperty("InputWorkspace", out_ws)
alg_fake.execute()
```
>You should see that if you uncomment the marked line, then the algorithm runs without an issue, else it will complain that there has been no name set for the output